### PR TITLE
feat(models): add Gemma 3 text decoder

### PR DESCRIPTION
**Pull request by hashkanna**
## Motivation

Fixes part of #1069.

This adds the first Gemma 3 support slice for native text-only serving. The goal of this PR is to land the Gemma 3 text decoder, checkpoint weight mapping, text-only `Gemma3ForConditionalGeneration` support, and focused correctness coverage before adding the multimodal vision/projector path in a follow-up PR.

This PR intentionally does not add image-input or multimodal image-token handling. Text-only Gemma 3 configs set `use_bidirectional_attention=False`, so the causal decoder path is correct for this slice; bidirectional/image-token handling is deferred to the multimodal follow-up.

## Modifications

- Add `python/sgl_jax/srt/models/gemma3.py` with:
  - `Gemma3ForCausalLM`
  - text-only `Gemma3ForConditionalGeneration`
  - `Gemma3Model`
  - `Gemma3DecoderLayer`
  - `Gemma3Attention`
  - `Gemma3MLP`

- Implement Gemma 3 text-decoder details:
  - `self_attn.q_norm` and `self_attn.k_norm`
  - explicit `head_dim` support
  - `query_pre_attn_scalar` attention scaling
  - partial rotary embedding support
  - sliding/full attention layer type handling
  - fallback from `sliding_window_pattern` when `layer_types` is not present
  - per-layer RoPE base selection for sliding vs full attention layers
  - Gemma 3 global/full-attention linear RoPE scaling
  - tied embedding/output head handling
  - HF-compatible embedding scaling semantics for token lookup vs provided `input_embedding`

- Add Gemma 3 weight mappings:
  - text checkpoint keys such as `model.layers.*`
  - q/k norm mappings
  - `language_model.*` prefixed mappings for conditional-generation checkpoints

- Add focused no-download unit tests for:
  - model registry resolution
  - sliding-window and RoPE helper behavior
  - global vs local RoPE scaling
  - `sliding_window_pattern` fallback
  - tiny model initialization
  - q/k norm weight mappings
  - conditional checkpoint prefix mappings
  - `forward_batch.input_embedding` mode gating

## Accuracy Tests

Local checks:

```bash
python -m compileall python/sgl_jax/srt/models/gemma3.py python/sgl_jax/test/test_gemma3_model.py
PYTHONPATH=python python -m pytest python/sgl_jax/test/test_gemma3_model.py -q
```

Result:

```text
13 passed, 13 warnings
```

Runtime validation on TPU v6e-1 with `google/gemma-3-270m`:

- Short prompt parity against HF Transformers:
  - prompt: `The capital of France is`
  - argmax matched HF: `Paris`
  - top-10 overlap: `10/10`

Sliding-window boundary parity against HF Transformers with hybrid SWA memory enabled:

| Input length | Argmax | Top-10 overlap | Max abs diff | Mean abs diff | P99 abs diff |
|---:|---|---:|---:|---:|---:|
| 128 | match | 10/10 | 0.25 | 0.048 | 0.156 |
| 511 | match | 10/10 | 0.625 | 0.098 | 0.375 |
| 512 | match | 10/10 | 0.625 | 0.202 | 0.375 |
| 513 | match | 10/10 | 0.328 | 0.059 | 0.195 |
| 710 | match | 10/10 | 0.562 | 0.205 | 0.406 |

The TPU run verified the expected Gemma 3 layer split for the 270M checkpoint:

```text
sliding layers: 0-4, 6-10, 12-16
full layers:    5, 11, 17
sliding_window_size: 512
hybrid SWA memory: enabled
```

End-to-end dataset validation:

| Model | Dataset | Prompting | Requests | Accuracy | Invalid | Reference |
|---|---|---:|---:|---:|---:|---|
| `google/gemma-3-270m` | GSM8K test | 5-shot | 1319 | 0.027 | 0.004 | No official same-checkpoint GSM8K reference published |
| `google/gemma-3-4b-pt` | GSM8K test | 8-shot | 1319 | 0.404 | 0.003 | Google Gemma 3 PT 4B: 0.384, 8-shot |

`google/gemma-3-4b-pt` launch command:

```bash
PYTHONPATH=python JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache python -u -m sgl_jax.launch_server \
  --model-path google/gemma-3-4b-pt \
  --host 0.0.0.0 \
  --port 30000 \
  --tp-size 1 \
  --device tpu \
  --dtype bfloat16 \
  --attention-backend fa \
  --page-size 64 \
  --mem-fraction-static 0.8 \
  --download-dir /tmp \
  --skip-server-warmup \
  --disable-radix-cache \
  --disable-precompile
```

`google/gemma-3-4b-pt` eval command:

```bash
PYTHONPATH=python python benchmark/gsm8k/bench_sglang_jax.py \
  --base-url http://127.0.0.1:30000 \
  --num-questions 1319 \
  --num-shots 8 \
  --parallel 64 \
  --result-file /tmp/gsm8k_4b_8shot_rope_results.jsonl \
  --output-file /tmp/gsm8k_4b_8shot_rope_outputs.txt
```

Result:

```text
Accuracy: 0.404
Invalid: 0.003
Latency: 285.387 s
Output throughput: 595.900 token/s
```

Reference numbers: Google/HF publish official same-checkpoint `Gemma 3 PT 270M` references for HellaSwag, BoolQ, PIQA, TriviaQA, ARC-c, ARC-e, and WinoGrande, but not GSM8K or MMLU for the 270M checkpoint. The official Google Gemma 3 model card lists `Gemma 3 PT 4B` GSM8K at 38.4% with 8-shot prompting, and this PR reaches 40.4% with the same checkpoint/protocol: https://ai.google.dev/gemma/docs/core/model_card_3

## Benchmarking and Profiling

No dedicated performance tuning is included in this PR.

This change adds model support and correctness coverage only. Performance tuning for Gemma 3 and multimodal serving are left for follow-up PRs.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [x] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by JamesBrianD** [comment]
Two things before this lands:

**Missing `AttentionType.DECODER_BIDIRECTIONAL`.** Upstream SGLang `gemma3_causal.py:228` sets `attn_type=AttentionType.DECODER_BIDIRECTIONAL` — Gemma3 prefill is bidirectional over the prompt, not causal. sgl-jax's `AttentionType` enum (`radix_attention.py:12-21`) only has `DECODER` and `ENCODER_ONLY`, so this PR silently falls back to causal. 270M short-prompt parity probably passes because the model is small enough that the difference gets lost in noise, or the HF baseline was also run causal. 

**The unit tests don't really test anything load-bearing.** Asserting registry lookup, helper return values, weight-mapping dict keys, shapes after `__init__` — none of this catches the actual risks (bidirectional gap, q/k norm misapplied, partial RoPE dim, per-layer rope_theta swap, residual ordering). Suggest dropping these. 

Concretely, in the PR description:

- Run MMLU or GSM8K against `google/gemma-3-270m` (and `gemma-3-4b-it` / `gemma-3-12b-it` if you have the compute)
- Paste the launch command (TP, dtype, attention backend, page size, …)
- Paste the eval command
- Paste the accuracy numbers
- Paste reference numbers from Google's Gemma3 tech report or the HF leaderboard for the same checkpoints

**Comment by hashkanna** [comment]
Thanks @JamesBrianD for the review!

This PR intentionally keeps the scope to the text-only causal decoder (`Gemma3ForCausalLM`) so the first Gemma 3 slice stays reviewable. From HF’s official Gemma 3 text configs, `use_bidirectional_attention=False`; the bidirectional behavior is needed for multimodal image-token blocks in `Gemma3ForConditionalGeneration`, which I’m planning to handle in the follow-up vision/multimodal PR.

For coverage, I added focused no-download behavioral tests. The PR now exercises SWA/full layer selection, text decoder attention type, Q/K norm before RoPE, per-layer RoPE/partial RoPE behavior, deferred decoder residual flow, and dual-prefix weight mapping.

I’m treating full MMLU/GSM8K as follow-up validation rather than in this first text-decoder PR, but I’m happy to run/add it here if you’d prefer that for review. Thanks!

**Comment by hashkanna** [comment]
Thanks @JamesBrianD, I added the e2e dataset validation to the PR description.

New coverage added there:

- TPU v6e-1 launch command for `google/gemma-3-270m` with TP=1, dtype=bf16, attention backend=`fa`, page size=64, and the exact serving flags used.
- GSM8K eval command using the repo's `/generate` benchmark harness.
- Full GSM8K test result: 1319 requests, 5-shot, accuracy `0.027`, invalid `0.004`.
- Reference note: the official Google/HF Gemma 3 cards publish same-checkpoint 270M references for HellaSwag/BoolQ/PIQA/TriviaQA/ARC/WinoGrande, but not GSM8K or MMLU for 270M; official GSM8K references are only published for larger Gemma 3 PT checkpoints, so I marked those as not directly comparable.

I also refreshed the local no-download test result in the description to `10 passed, 13 warnings`.

**Comment by hashkanna** [comment]
Thanks for the review. I pushed an update and refreshed the PR description with the requested e2e dataset result.

Summary of the update:
- Added text-only `Gemma3ForConditionalGeneration` support with `language_model.*` weight mappings.
- Fixed Gemma 3 global/full-attention RoPE linear scaling; local/sliding RoPE remains unscaled to match HF behavior.
- Kept multimodal/image-token handling explicitly out of scope for the follow-up PR.

Validation:
- `PYTHONPATH=python python -m pytest python/sgl_jax/test/test_gemma3_model.py -q` -> `13 passed`
- `google/gemma-3-4b-pt` GSM8K 8-shot on TPU v6e-1, `bfloat16`, FA backend, page size 64:
  - accuracy: `40.4%`
  - invalid: `0.3%`
  - official Google Gemma 3 PT 4B GSM8K 8-shot reference: `38.4%`

Also confirming the bidirectional attention point: text-only Gemma 3 configs set `use_bidirectional_attention=False`, so this PR keeps the causal decoder path. Bidirectional/image-token handling is deferred to the multimodal follow-up.
